### DISCUSSION TITLE
mosquitto_pub --help - example

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#16209] [Q&A]
I know how to control my device using the browser and curl but have yet to figure out the client.
After looking at countless articles, I still cannot find one example that toggles the relays I have.

I know the MQTT server is working because I can echo things to it.

In one terminal, I run the following command;
$ mosquitto_pub -d -t garden -m "{"Power1" : "on" , "cmd2" : "on"}"
Client null sending CONNECT
Client null received CONNACK (0)
Client null sending PUBLISH (d0, q0, r0, m1, 'garden', ... (25 bytes))
Client null sending DISCONNECT

On another, I can see the server messages;
{Power1 : on , cmd2 : on}

Looking online to control the relays, nothing works.

mosquitto_pub -h 127.0.0.1 -t garden -m on
mosquitto_pub -h 127.0.0.1 -t garden/cmd/Power1 -m on
mosquitto_pub -t garden/cmd/Power1 -m on
mosquitto_pub -h 192.168.1.134 -t garden/cmd/Power1 -m on
mosquitto_pub -h 127.0.0.1 -t garden/cmnd/Power1%20TOGGLE
mosquitto_pub -h 127.0.0.1 -t garden//cm?cmnd=Power%20TOGGLE
mosquitto_pub -h localhost -t '/hello' -m "{"cmd1" : "on" , "cmd2" : "on"}"

And a very long list of more commands found on the net that don't work.
Can someone suggest at least one command that might work?

The device is configured with the MQTT server but nothing ever shows up in the server messages.
The device is this one; https://www.aliexpress.com/item/3256802966722010.html

I feel bad having to ask but I'm at a loss now and need a lead of some kind.

**Comment by barbudor** [comment]
Unless you changed the fulltopic, you get order of prefix wrong
`mosquitto_pub -h 192.168.1.134 -t cmnd/garden/Power1 -m on`

Also assuming that if you are using mosquitto 2.x you have explicitly enabled anonymous connection (it was implicit in 1.6.x)

**Reply by dac2020** [reply]
Hi,

I installed a ubuntu server with mosquitto.
mosquitto_sub version 2.0.14 running on libmosquitto 2.0.14.
It's a default install so unless it has a default password, then I didn't add one. 
Looking at the config, it's very minor so all default.

$ mosquitto_pub -h 192.168.1.134 -t cmnd/garden/Power1 -m on
Error: Connection refused
$ mosquitto_pub -h 127.0.0.1 -t cmnd/garden/Power1 -m on

Nothing shows in the MQTT messages.

$ mosquitto_sub -h localhost -t garden

**Reply by sfromis** [reply]
If `192.168.1.134` is your Tasmota device, that's not what you should do. The `-h` parameter should point at the host where the broker is running. If you actually are trying to point at the right host, the problem may be the next question:

Did you explicitly enable anonymous connections in `mosquitto.conf`, **and** allow for network connections? You have to do both, alternatively set up to use passwords (which I'd not recommend until later, or likely never).

Here's my `mosquitto.conf` doing both required steps:
```
listener 1883 0.0.0.0
allow_anonymous true
```

**Comment by dac2020** [comment]
Correct. I didn't want to add complexity by adding network limits or passwords so I left it alone.
I added your directives above but still no control.

No firewall is active on the server.
Do the directives have to be in a specific section of the conf file? I saw one example where there were sections but wasn't able to find it again yet.

192.168.1.134 / 127.0.0.1 is the broker host.
The device is at 192.168.254.201 and has full access.

**Reply by dac2020** [reply]
I do see the device in the broker.

'''
$ mosquitto_sub -v -t '#'
tele/garden/LWT Online
tasmota/discovery/7821846D54E0/config {"ip":"192.168.254.201","dn":"4chRelay-ch-01","fn":["4chRelay-ch-01","4chRelay-ch-01","4chRelay-ch-01                                                                                   ","4chRelay-ch-01",null,null,null,null],"hn":"garden-5344","mac":"7821846D54E0","md":"ESP32-DevKit","ty":0,"if":0,"ofln":"Offline","onln":"                                                                                   Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"12.0.2","t":"garden","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,1,1,1,                                                                                   0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1                                                                                   ],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,                                                                                   null,null],"btn":[0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":1,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"s                                                                                   ho":[0,0,0,0],"sht":[[0,0,0],[0,0,0],[0,0,0],[0,0,0]],"ver":1}
tasmota/discovery/7821846D54E0/sensors {"sn":{"Time":"1970-01-01T22:39:35","ESP32":{"Temperature":32.8},"TempUnit":"C"},"ver":1}
'''

**Reply by dac2020** [reply]
Maybe it just took a minute for some reason but now it works.

mosquitto_pub -h 127.0.0.1 -t cmnd/garden/Power1 -m on

Response;
cmnd/garden/Power1 on
stat/garden/RESULT {"POWER1":"ON"}
stat/garden/POWER1 ON

**Reply by dac2020** [reply]
The thing I don't get is where to find information like what else can I get from the device such as status for example. 
The commands I assume are somewhat similar across different devices but how do you find out what those are?

**Comment by dac2020** [comment]
Ah, I think I see how now. Once the device is able to communicate with the broker, it shows you what it can do.


POWER1":"OFF
POWER2":"OFF
POWER3":"OFF
POWER4":"OFF
etc. It's in the JSON string being sent.

But, it is not clear how you use that.

mosquitto_pub -h 127.0.0.1 -t cmnd/garden/Temperature
Nope.

For example, I'd like to change the temp from C to F.
I can see it's called; TempUnit":"C"

% mosquitto_pub -h 127.0.0.1 -t cmnd/garden/TempUnit -m F

Result;
cmnd/garden/TempUnit F
stat/garden/RESULT {"Command":"Unknown"}

Nope.

No problem, now I can search the net again :).

**Reply by sfromis** [reply]
No, what you see in those payloads are not commands, it is simply Tasmota publishing details. Some of those details are published with a field name similar to a command name, but this is in no way a rule.

This page documents available commands: https://tasmota.github.io/docs/Commands/
All the commands can be used both via console, MQTT, serial, http, rules and Berry.
On that page, you can see things like the command to select Fahrenheit vs Celsius is `SetOption8`

Note: For reliable operation, you have to give Tasmota a real time source, typically a NTP server. Default is to use public NTP servers on the Internet, but you can also setup a local NTP server if you prefer.

**Reply by dac2020** [reply]
I understand now. I'll check that page.
Each step is both frustrating and exciting at the same time thank to your help and others.
